src/turn-command-manager.ts

```diff
@@ -0,0 +1,477 @@
+import { MoveHeaderAttr } from "./data/move-attrs/move-header-attr";
+import type { Pokemon } from "./field/pokemon";
+import { PokemonMove } from "./field/pokemon-move";
+import { globalScene } from "./global-scene";
+import { BypassSpeedChanceModifier } from "./modifier/modifier";
+import { AttemptCapturePhase } from "./phases/attempt-capture-phase";
+import { AttemptRunPhase } from "./phases/attempt-run-phase";
+import { MoveHeaderPhase } from "./phases/move-header-phase";
+import { MovePhase } from "./phases/move-phase";
+import { SwitchSummonPhase } from "./phases/switch-summon-phase";
+import { BooleanHolder, isNullOrUndefined, randSeedShuffle } from "./utils";
+
+/**
+ * Interface representing an action taken by a Pokemon for the turn.
+ * Encompasses both Player and Enemy commands.
+ */
+export interface TurnCommand {
+  /** The {@linkcode Pokemon} carrying out the action */
+  pokemon: Pokemon;
+  /** The type of action to carry out */
+  command: BattleCommand;
+  /**
+   * The cursor index given by the user.
+   * Used for {@linkcode Command.POKEMON | switch commands}.
+   */
+  cursor?: number;
+  /**
+   * The move for the Pokemon to use.
+   * Used for {@linkcode Command.FIGHT | fight commands}.
+   */
+  move?: QueuedMove;
+  /**
+   * The {@linkcode Pokemon} to target with the action.
+   * Used for {@linkcode Command.FIGHT | fight} and
+   * {@linkcode Command.BALL | ball} commands
+   */
+  targets?: BattlerIndex[];
+  /** Any other arguments given with this command */
+  args?: any[];
+}
+
+export class TurnCommandManager {
+  /**
+   * The internal {@linkcode TurnCommand} queue.
+   *
+   * NOTE: This is only `public` to facilitate unit tests that override turn order.
+   * Please use this class's API to access and modify turn commands instead
+   * of accessing this array directly.
+   */
+  public turnCommands: TurnCommand[];
+  private orderIndex: number;
+
+  constructor() {
+    this.turnCommands = [];
+    this.orderIndex = 0;
+  }
+
+  // --------------  BEGIN PUBLIC METHODS  -------------- //
+
+  /**
+   * @returns the {@linkcode TurnCommand} in the turn sequence
+   * for the given {@linkcode Pokemon}.
+   */
+  public getCommand(pokemon: Pokemon): TurnCommand | undefined {
+    return this.turnCommands.find((tc) => tc.pokemon === pokemon);
+  }
+
+  /**
+   * Adds a command to the command queue.
+   * After this is called, turn order should be reset
+   * using {@linkcode setTurnOrder}.
+   * @param turnCommand the command to add
+   */
+  public addCommand(turnCommand: TurnCommand) {
+    const { pokemon } = turnCommand;
+    if (pokemon.turnData) {
+      pokemon.turnData.turnCommand = turnCommand;
+    }
+    this.turnCommands.push(turnCommand);
+  }
+
+  /**
+   * Sorts the turn command queue by the command's turn order
+   * @param quiet if `true`, applies abilities and other field effects silently
+   */
+  public setTurnOrder(quiet: boolean = true): void {
+    this.shuffle();
```

Comment:
```suggestion
    this.shuffle(); // shuffle the list before sorting so speed ties produce random results
```
adding back the command from startTurnPhase cause I was confused at first why this was being done